src/Front/DesugarLists.hs

```diff
@@ -1,29 +1,52 @@
 -- | Desugar ListExpr nodes into App nodes
 module Front.DesugarLists
   ( desugarLists
-  ) where
+  )
+where
 
-import qualified Common.Compiler               as Compiler
-import           Front.Ast                      ( Definition(..)
-                                                , Expr(..)
-                                                , Program(..)
-                                                , TopDef(..)
-                                                )
+import Data.Generics
+import qualified Common.Compiler as Compiler
+import Front.Ast (Definition(..), Expr(..), Program(..), TopDef(..), Literal(..))
+import Common.Identifiers
 
 -- | Desugar ListExpr nodes inside of an AST 'Program'.
 desugarLists :: Program -> Compiler.Pass Program
-desugarLists (Program decls) = return $ Program $ desugarTop <$> decls
- where
-  desugarTop (TopDef d) = TopDef $ desugarDef d
-  desugarTop t          = t
+desugarLists (Program decls) = return $ Program $ desugarTop <$> decls 
+  where
+    desugarTop (TopDef d) = TopDef $ desugarDef d
+    desugarTop t = t
+    desugarDef (DefFn v bs t e) = DefFn v bs t $ everywhere (mkT desugarExpr) e
+    desugarDef (DefPat b e    ) = DefPat b $ everywhere (mkT desugarExpr) e
 
-  desugarDef (DefFn v bs t e) = DefFn v bs t $ desugarExpr e
-  desugarDef (DefPat b e    ) = DefPat b $ desugarExpr e
 
--- | Transform a node of type ListExpr into a node of type App 
+{-
+appidhelper:: A.ListExpr -> A.ListExpr
+appidhelper [] = [(A.Id (I.Identifier "App")) (A.Id (I.Identifier "Cons")) (A.Id (I.Identifier "Nil"))]
+appidhelper (h:t) = A.Id (I.Identifier "App") (A.Id (I.Identifier "Cons")) h : appidhelper t
+
+ListExpr is not a type, it's a data constructor.
+Data constructors return an instance of a type. 
+ListExpr is a data constructor that returns an instance of type Expr.
+It's not possible to have a function that takes in
+something of type ListExpr and returns something of type ListExpr, because a ListExpr is not a type.
+It is possible to have a function that takes in 
+something of type Expr and returns something of type Expr!
+Then you can special case on instance of ListExpr.
+I think this is what you wanted to do.
+-}
+
+appidhelper:: [Expr] -> Expr
```

Comment:
could you rename this function something shorter? Maybe just call it helper? Also, could you make this a local function by putting it in a `where` clause? For ex., something like 
```
desugarExpr :: Expr -> Expr
desugarExpr (ListExpr es) = appidhelper es
  where helper:: A.ListExpr -> A.ListExpr
        helper [] = (Id nil)
        ...
desugarExpr e = e
```